### PR TITLE
Test driven

### PR DESCRIPTION
**Pull request by kev1n80**

### CODE REVIEW

File: walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java [hunk 1]
```diff
@@ -30,4 +30,24 @@ public void testGreeting() {
...
+
+    // Whitespace should be trimmed
+    Assert.assertEquals("Hello Ada", greeting);
+  }
+
+  @Test
+  public void testGreetingTrimsSymbols() {
+    Greeter greeter = new Greeter();
+
+    String greeting = greeter.greet("  @#$ Ada   %");
```
Comment (saucepwn) [inline walkthroughs/week-5-tdd/intro/src/test/java/com/google/sps/GreeterTest.java:48]: Good job covering extra corner cases

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,198 @@
...
+   * @return An ArrayList that contains an array of start and end times, and 
+   *     will not have any events with the same start time.
+   */
+  public static ArrayList<int[]> eventToTime(Event[] eventsArray) 
+      throws Exception {
+    ArrayList<int[]> eventTimes = new ArrayList<>();
+    TimeRange previousEventTR = TimeRange.fromStartDuration(0, 0);
+
+    for (Event event : eventsArray) {
+      TimeRange eventTR = event.getWhen();
```
Comment (saucepwn) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: nit: Minor edit, but for readability I'd expand `eventTR` to `eventTimeRange` and also expand TR to TimeRange in other local variables too.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,198 @@
...
+          eventTimes.add(eventTime);
+          previousEventTR = eventTR;
+        }
+      } else if (eventTRStart == previousEventTRStart) {
+        // if they are the same, keep the one with the longer duration
+        if (eventTR.duration() > previousEventTR.duration()) {
+          int lastIndex = eventTimes.size() - 1;
+          if (eventTimes.size() == 0) {
+            lastIndex = 0;
+          }
```
Comment (saucepwn) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: saves ya 3 lines

```suggestion
          int lastIndex = Math.max(eventTimes.size() - 1, 0);
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,198 @@
...
+    ArrayList<Event> filteredEvents = new ArrayList<Event>();
+    for (Event event : events) {
+      // if the events share an attenddee then add it to the array
+      if (pred.test(event)) {
+        filteredEvents.add(event);
+      }
+    }
+    Event[] filteredEventsArray = new Event[filteredEvents.size()];
+    filteredEventsArray = filteredEvents.toArray(filteredEventsArray);
+    return filteredEventsArray;
```
Comment (saucepwn) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]:
```suggestion
    return filteredEvents.toArray(filteredEventsArray);
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,198 @@
...
+   * Returns all possible time periods throughout the day when everybody 
+   *    attending this meeting is available. 
+   * Time Complexity: O(n^2 * ln(n))
+   * 
+   * @param events All events that are occurring  
+   * @param request The meeting that the user wants to create and find time for 
+   * @return an array of TimeRange objects
+   */
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) throws Exception{
+    // throw new UnsupportedOperationException("TODO: Implement this method.");
```
Comment (saucepwn) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: remove

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,198 @@
...
+
+    // check if duration of meeting is longer than a day or a negative number
+    int durationMeeting = Math.toIntExact(request.getDuration());
+    // long durationMeeting = request.getDuration();
```
Comment (saucepwn) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: remove

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,198 @@
...
+   * 
+   * @param events All events that are occurring  
+   * @param request The meeting that the user wants to create and find time for 
+   * @return an array of TimeRange objects
+   */
+  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) throws Exception{
+    // throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // check if duration of meeting is longer than a day or a negative number
+    int durationMeeting = Math.toIntExact(request.getDuration());
```
Comment (saucepwn) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: What unit is this duration in? Add that to the variable name. For example, if it's minutes, rename: `durationMeetingMinutes` etc.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,198 @@
...
+    // Get the times of the event and remove events with the same times
+    //    (but keep the longest duration) and the events that are contained in 
+    //    another event
+    ArrayList<int[]> times = new ArrayList<int[]>();
+    try {
+      times = eventToTime(eventsArray);  
+    } catch (Exception e) {
+      String errorMessage = "Error: " + e.getMessage();
+      System.err.println(errorMessage);
+      throw e;
```
Comment (saucepwn) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:194]: This shouldn't be necessary as the exception will be printed to the console as it bubbles up to the top level

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/comparator/SortEventsByTime.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+public class SortEventsByTime implements Comparator<Event> { 
+  /** 
+  * Used for sorting in ascending order of start time and for a tie breaker
+  *     in descending order of duration.
+  * Time complexity: O(1)
+  *
+  * @param first the first Event
+  * @param second the second Event
+  * @return an int that states the ordering of the two events
+  */
```
Comment (saucepwn) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/comparator/SortEventsByTime.java]: Indent these lines by +1 space

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/comparator/SortEventsByTime.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  @Override
+  public int compare(Event first, Event second) { 
+    TimeRange firstTime = first.getWhen();
+    TimeRange secondTime = second.getWhen();
+    int order = TimeRange.ORDER_BY_START.compare(firstTime, secondTime);
+    /** 
+    * if they start at the same time, the one with the longest duration 
+    *     shows up first
+    */
```
Comment (saucepwn) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/comparator/SortEventsByTime.java]: Indent these lines by +1 space

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/comparator/SortEventsByTime.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+    if (order == 0) {
+      return - Long.compare(firstTime.duration(), secondTime.duration());
```
Comment (saucepwn) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/comparator/SortEventsByTime.java]: Does this complie with the - symbol?

**Review by saucepwn — APPROVED** [review]
